zombienet_tests/functional/0003-beefy-and-mmr.zndsl

```diff
@@ -3,45 +3,37 @@ Network: ./0003-beefy-and-mmr.toml
+validator: reports substrate_beefy_validator_set_id is 0
+validator-unstable: reports substrate_beefy_validator_set_id is 0
 
 # Verify voting happens and 1st mandatory block is finalized within 1st session.
-validator-0: reports substrate_beefy_best_block is at least 1 within 60 seconds
-validator-1: reports substrate_beefy_best_block is at least 1 within 60 seconds
-validator-2: reports substrate_beefy_best_block is at least 1 within 60 seconds
-validator-3: reports substrate_beefy_best_block is at least 1 within 60 seconds
+validator: reports substrate_beefy_best_block is at least 1 within 60 seconds
+validator-unstable: reports substrate_beefy_best_block is at least 1 within 60 seconds
 
-# Pause validator-3 and test chain is making progress without it.
-validator-3: pause
+# Pause validator-unstable and test chain is making progress without it.
+validator-unstable: pause
 
 # Verify validator sets get changed on new sessions.
-validator-0: reports substrate_beefy_validator_set_id is at least 1 within 70 seconds
-validator-1: reports substrate_beefy_validator_set_id is at least 1 within 70 seconds
-validator-2: reports substrate_beefy_validator_set_id is at least 1 within 70 seconds
+validator: reports substrate_beefy_validator_set_id is at least 1 within 70 seconds
 # Check next session too.
-validator-0: reports substrate_beefy_validator_set_id is at least 2 within 130 seconds
-validator-1: reports substrate_beefy_validator_set_id is at least 2 within 130 seconds
-validator-2: reports substrate_beefy_validator_set_id is at least 2 within 130 seconds
+validator: reports substrate_beefy_validator_set_id is at least 2 within 130 seconds
 
 # Verify voting happens and blocks are being finalized for new sessions too:
 # since we verified we're at least in the 3rd session, verify BEEFY finalized mandatory #21.
-validator-0: reports substrate_beefy_best_block is at least 21 within 130 seconds
-validator-1: reports substrate_beefy_best_block is at least 21 within 130 seconds
-validator-2: reports substrate_beefy_best_block is at least 21 within 130 seconds
+validator: reports substrate_beefy_best_block is at least 21 within 130 seconds
 
-# TODO (issue #11972): Custom JS to test BEEFY RPCs
-# TODO (issue #11972): Custom JS to test MMR RPCs
+# Custom JS to test BEEFY RPCs.
+validator-0: js-script ./0003-beefy-finalized-heads.js with "validator-0,validator-1,validator-2" return is 1 within 5 seconds
 
-# Resume validator-3 and verify it imports all BEEFY justification and catches up.
-validator-3: resume
-validator-3: reports substrate_beefy_validator_set_id is at least 2 within 30 seconds
-validator-3: reports substrate_beefy_best_block is at least 21 within 30 seconds
+# Custom JS to test MMR RPCs.
+validator: js-script ./0003-mmr-leaves.js with "21" return is 1 within 5 seconds
+validator: js-script ./0003-mmr-generate-and-verify-proof.js with "validator-0,validator-1,validator-2" return is 1 within 5 seconds
```

Comment:
Nit: If I understand correctly, now this will be called for each validator in group `validator` with arguments `validator-0,validator-1,validator-2` (because of `validator: ...`). So it will do the same check 3 times. Should we drop `with "validator-0,validator-1,validator-2"` ? Or use `validator-0: js-script ./0003-mmr-generate-and-verify-proof.js with "validator-0,validator-1,validator-2" return is 1 within 5 seconds` ?

Reply:
You're right that it will check this for every validator. The intent here is full coverage: every validator generates the proof, and we check that all other validators verify it successfully. Having this full coverage is inexpensive: ~250ms per run, so overhead of ~0.5s against only checking a single one, vs. overall runtime of ~150s of `0003-beefy-and-mmr.zndsl` on my system.  

For the `beefy-finalized-heads` test, I've cut it down to `validator-0` only (https://github.com/paritytech/polkadot/commit/89ced7dd934d66a6eebdfab3abcff3c47db010ac), since here there's indeed no additional coverage from running the test on every validator - the validator arg isn't even used.

Reply:
Oh ok, makes sense.